LDMP/models/datasets.py

```diff
@@ -178,8 +215,80 @@ def dump(self) -> str:
                 f, default=json_serial, sort_keys=True, indent=4, separators=(',', ': ')
             )
         self.dumped.emit(descriptor_file_name)
+
+        self.__fileName = descriptor_file_name
         return descriptor_file_name
 
+    def fileName(self):
+        return self.__fileName
+
+    @staticmethod
+    def datetimeRepr(dt: datetime) -> str:
+        return datetime.strftime(dt, '%Y/%m/%d (%H:%M)')
+
+    @staticmethod
+    def toDatetime(dt: str, fmt: str = None) -> datetime:
+        if fmt is None:
+            return datetime.fromisoformat(dt)
+        return datetime.strptime(dt, fmt)
+
+    def download(self, datasets_refresh: bool = True) -> None:
+        """Download Dataset related to a specified Job in a programmatically defined filename and folder.
+        Because a new dataset JSON descriptor is created => Datasets sync
+
+        Args:
+            datasets_refresh (bool): if have to trigger Datasets().sync() after downloaded.
+
+        Returns:
+            None
+        """
+        res = Jobs().jobById(str(self.run_id)) # casting because can be UUID
+        if res is None:
+            log(tr('No Job available in cache with run id: {}').format(str(self.run_id)))
+            return
+        job_filename, job = res # unwrap tuple
+
+        # get path of the current dataset descriptos
+        res = Datasets().datasetById(str(self.run_id)) # casting because can be UUID
+        if res is None:
+            log(tr('No Dataset available in cache with run id: {}').format(str(self.run_id)))
+            return
+        dataset_filename = res[0]
+
+        # Check if we need a download filename - some tasks don't need to 
+        # save data, but if any of the chosen tasks do, then we need to 
+        # choose a folder. Right now only TimeSeriesTable doesn't need a 
+        # filename.
+        base_dir = os.path.dirname(dataset_filename)
+        f = None
+        if job.results.get('type') != 'TimeSeriesTable':
+            # create result filename basing on:
+            # run_id + scriptName + task_name (if any)
+            fileBaseName = None
+            if job.taskName:
+                fileBaseName = u'{}_{}_{}'.format(job.runId, job.scriptName, job.taskName)
+            else:
+                fileBaseName = u'{}_{}_{}'.format(job.runId, job.scriptName)
+            
+            # set the folder where to save result
+            f = os.path.join(base_dir, fileBaseName)
+            log(u"Downloading results to {} with basename {}".format(os.path.dirname(f), os.path.basename(f)))
+
+        log(u"Processing job {}.".format(job.runId))
+        result_type = job.results.get('type')
+        if result_type == 'CloudResults':
+            download_cloud_results(job.raw, f, self.tr, add_to_map=False)
+            self.downloaded.emit(f)
+            if datasets_refresh:
+                Datasets().sync()
```

Comment:
shouldn't we make this a static method?

Reply:
could be static because it's a singleton working on class variable... btw I would avoid to access it without a single instance on which to work. e.g I would leave it as singleton to force user use singleton constructor and explicitly dige the idea they are working on a single instance